plan/build_key_info.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+)
+
+// Columns of an aggregation's schema is an bijection with aggregation's aggFuncs.
+// Sometimes we need to convert from one side to another, then this function is needed.
+func (p *Aggregation) buildSchemaByAggFuncs() expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, p.schema.Len()))
+	for _, fun := range p.AggFuncs {
+		if col, isCol := fun.GetArgs()[0].(*expression.Column); isCol && fun.GetName() == ast.AggFuncFirstRow {
+			schema.Append(col)
+		} else {
+			// If the arg is not a column, we add a column to occupy the position.
+			schema.Append(&expression.Column{
+				Position: -1})
+		}
+	}
+	return schema
+}
+
+func (p *Aggregation) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	// Dealing with p.AggFuncs.
+	schemaByFuncs := p.buildSchemaByAggFuncs()
+	for _, key := range p.GetChildren()[0].GetSchema().Keys {
+		indices := schemaByFuncs.GetColumnsIndices(key)
+		if indices == nil {
+			continue
+		}
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, i := range indices {
+			newKey = append(newKey, p.schema.Columns[i])
+		}
+		p.schema.Keys = append(p.schema.Keys, newKey)
+	}
+	// Dealing with p.GroupbyCols
+	// This is only used for optimization and needn't to be pushed up, so only one is enough.
+	schemaByGroupby := expression.NewSchema(p.groupByCols)
+	for _, key := range p.GetChildren()[0].GetSchema().Keys {
+		indices := schemaByGroupby.GetColumnsIndices(key)
+		if indices == nil {
+			continue
+		}
+		newKey := make([]*expression.Column, 0, len(key))
+		for _, i := range indices {
+			newKey = append(newKey, schemaByGroupby.Columns[i])
```

Comment:
GroupByCols may not appear in p.schema.Columns.
Such as `select sum(b) from t group by a`, schema will only contain one row corresponded to `sum(b)`.